src/vs/workbench/contrib/terminal/browser/terminalProcessManager.ts

```diff
@@ -328,7 +333,20 @@ export class TerminalProcessManager extends Disposable implements ITerminalProce
 		const useConpty = this._configHelper.config.windowsEnableConpty && !isScreenReaderModeEnabled;
 		const shouldPersist = this._configHelper.config.enablePersistentSessions && !shellLaunchConfig.isFeatureTerminal;
 
-		this._terminalInstanceService.onPtyHostUnresponsive(() => this._onPtyDisconnect.fire());
+		this._register(this._terminalInstanceService.onPtyHostUnresponsive(() => {
+			this.isDisconnected = true;
+			this._onPtyDisconnect.fire();
+		}));
+		this._ptyResponsiveListener = this._terminalInstanceService.onPtyHostResponsive(() => {
```

Comment:
maybe a comment like 
once the pty host has a heartbeat, (when does this happen, do we always restart the ptyHost on disconnect?) reconnect the disconnected ptys

Reply:
Added comments in main